Furo docs theme with styling

asmeurer opened the pull request:
#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

This is based off of https://github.com/sympy/sympy/pull/22864, but I've updated the styling of the Furo theme. This is not at all finished yet, and we also haven't formally decided to switch to Furo yet either. 

Demo site here: https://www.asmeurer.com/sympy-furo-demo/


#### Brief description of what is fixed or changed

I've done a lot of styling, but there are still things that can be improved. Please build the docs and play around with it. If you find something that is low contrast or the wrong color, it probably means it is something I didn't notice.

Also please try both the light and dark theme (you can click the icon at the top of the page).

I haven't changed any fonts or font sizes, except for the code examples. It's a possibility to do this, but I'd like feedback. 

CC @bertiewooster 

#### Other comments

**TODOs**

- [x] Check if fonts should be adjusted (different fonts, sizes, etc.)
- [x] Add a version selector to the sidebar
- [x] Adjust the logo size on mobile (or alternately, make the whole sidebar scrollable)
- [x] Fix all remaining contrast issues
- [x] Remove all text referencing SymPy Live

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- other
  - The documentation now uses the [Furo](https://pradyunsg.me/furo/) Sphinx theme. 
<!-- END RELEASE NOTES -->

bertiewooster committed:
Using theme furo

bertiewooster committed:
Added furo to doc/requirements.txt
Build works to a point,
then gives error:
Extension error (furo):
Handler <function _html_page_context at 0x7f674f1f6a60> for event 'html-page-context' threw an exception (exception: '_static/styles/furo.css' is not in list)
make: *** [Makefile:60: html] Error 2

bertiewooster committed:
Commented out html_style to build Furo

bertiewooster committed:
Removed default.css_t to disable custom styling
In conf.py, set language = 'en' to try to resolve accessibility issue:
<html> element does not have a [lang] attribute

bertiewooster committed:
Disable SymPy Live

bertiewooster committed:
Undo adding furo to release/requirements.txt

asmeurer committed:
Restyle the Furo theme

This uses the SymPy colors and more closely matches what we had before.

There are still some things that can be done here, and I have probably missed
some things that don't look good, especially in dark mode.

asmeurer committed:
Some more fixes to the docs theming

asmeurer commented:
Note: we should probably rebase away the submodule addition before merging any of this.

asmeurer commented:
I've uploaded a demo site so you don't have to build this yourself https://github.com/asmeurer/sympy-furo-demo

bertiewooster commented:
Thanks @asmeurer. To save folks a click, the built site is at https://www.asmeurer.com/sympy-furo-demo/

bertiewooster commented:
Great start! Comments:

1. The Search box appears to connect to the body of the page, because they're about the same color. The Furo example site has the search box the same color as the rest of the left navbar. Is that possible and advisable? I can imagine that having a dark green background would make it hard to read black text being input; is it possible to make the input text a light color?
2. [Lighthouse reports](https://googlechrome.github.io/lighthouse/viewer/?psiurl=https%3A%2F%2Fwww.asmeurer.com%2Fsympy-furo-demo%2Fmodules%2Fcore.html%23module-sympy.core.random&strategy=mobile&category=performance&category=accessibility&category=best-practices&category=seo&category=pwa&utm_source=lh-chrome-ext) that the accessibility is very good, at 98/100. The one problem it found is `Background and foreground colors do not have a sufficient contrast ratio.`
    1. One situation where this occurs is the light green text in a function template light gray box, for example [`sympy.core.sympify.`](https://www.asmeurer.com/sympy-furo-demo/modules/core.html#id1). It also occurs for links (for example, "[next section](https://www.asmeurer.com/sympy-furo-demo/tutorial/gotchas.html#tutorial-gotchas-symbols)"), which are light green on white background. Perhaps make those types of text dark green instead of light? Personally, I prefer lighter colors for large page elements, so maybe make the left navbar a light green, and the text in the function template a dark green?
    2. Another is the [text in a code block](https://www.asmeurer.com/sympy-furo-demo/tutorial/intro.html#what-is-symbolic-computation). Perhaps make the code block background color lighter?

asmeurer commented:
> The Search box appears to connect to the body of the page, because they're about the same color. The [Furo example site](https://sphinx-themes.org/sample-sites/furo/) has the search box the same color as the rest of the left navbar. Is that possible and advisable? I can imagine that having a dark green background would make it hard to read black text being input; is it possible to make the input text a light color?

Sure that's doable. I'm not sure how good it would look. Having it blend in does make it harder to see. I guess the flip side of that is that it's less distracting. The current docs site has the search bar white https://docs.sympy.org/dev/index.html.

> [Lighthouse reports](https://googlechrome.github.io/lighthouse/viewer/?psiurl=https%3A%2F%2Fwww.asmeurer.com%2Fsympy-furo-demo%2Fmodules%2Fcore.html%23module-sympy.core.random&strategy=mobile&category=performance&category=accessibility&category=best-practices&category=seo&category=pwa&utm_source=lh-chrome-ext) that the accessibility is very good, at 98/100. The one problem it found is Background and foreground colors do not have a sufficient contrast ratio.

Yeah, I noticed some of these things as well (the Chrome inspect accessibility thing flagged them). 

I'm a little unsure how to fix some of them. A problem I noticed with these color contrast checkers is that they only consider the contrast of the text against its background. They don't consider the contrast of the text with the surrounding text. This is particularly important for the link color. We can make the (currently green) links have a better contrast against the background by making them darker, but then they become harder to see as links against the surrounding black text (you can see what I mean if you right click "inspect" a link in Chrome and then click the color in the inspector and choose the suggested higher contrast color).

I can try flipping the dark and light greens to see how it looks. 

I was also planning on adjusting the code block background (which should be easier than trying to restyle all the different syntax highlighting colors). The problem is the Chrome dev tools only support suggesting foreground colors, not background colors. But I did find https://accessible-colors.com/ which allows both. AAA is hard to achieve while still keeping distinguishable colors, but we should be able to get AA I think. I also tried just messing with the color picker, but I had a hard time finding something I liked aesthetically. I do personally like how the code blocks stand out as I think the examples are the most important part of documentation.

pradyunsg commented on doc/src/conf.py:
```diff
@@ -142,7 +142,52 @@
+
+html_theme_options = {
+    "light_css_variables": css_settings_common,
+    "dark_css_variables": {**css_settings_common,
```
You shouldn't need this -- the light mode variables get inherited for dark mode.

pradyunsg replied:
See https://pradyunsg.me/furo/customisation/colors/#how-light-and-dark-mode-work for the relevant details.

asmeurer replied:
I didn't realize this. It would be nice if this were mentioned at https://pradyunsg.me/furo/customisation/#light-css-variables-dark-css-variables

asmeurer replied:
This was working, but I updated to the latest Furo and it stopped working and I had to separate them again.

pradyunsg commented:
> Add a version selector to the sidebar

Relevant here: https://pradyunsg.me/furo/customisation/sidebar/. See also https://github.com/pradyunsg/furo/issues/372.

moorepants commented:
This is looking pretty nice. I like the green!
I noticed that the API rendering isn't so great:
![image](https://user-images.githubusercontent.com/276007/155696302-5c3b1fef-a04c-4cb4-b3ca-4c4ffeaf6caa.png)

The name of the function/class/method doesn't stand out. The bold "parameters" "returns" etc overwhelms the name of the function. The name of the functions should be the most distinct thing on the page. Also the notes are way too strong (in color I guess):
![image](https://user-images.githubusercontent.com/276007/155696480-955e0f2c-8424-48d0-9cf4-52d1e28ad277.png)

This also detracts from the function name which should be the most prominent thing.

moorepants commented:
If you scroll through the api listing, the function names should be the only thing you can detect while scrolling (or the primary thing).

namannimmo10 commented:
Thanks for this.  The new docs site looks really nice!

asmeurer committed:
Move the Furo custom CSS variables back into html_theme_options

The Furo "dark" CSS variables automatically inherit the light variable values.

asmeurer committed:
Disable the fancy scrolling behavior in the Furo theme

The scrolling makes jumping to sections on very large pages too slow.

See https://github.com/pradyunsg/furo/discussions/384.

asmeurer committed:
Make the headings in the API docs easier to see

asmeurer committed:
Reduce the blockquote padding

The API docs render the parameter descriptions as blockquotes. We removed all
other blockquote styling, but the vertical padding was still there, which was
too much and made it harder to read.

asmeurer committed:
Some more improvements to the documentation styling:

- Make the "see also" and "notes" headers gray instead of green. This will make
  them stand out on the page less.
- Make the "class", "property" and so on black instead of green. This makes it
  easier to see the function name itself.
- Make the function name headers larger.

asmeurer committed:
Use a slightly less bold font for the API doc subheaders ("Examples" and so on)

asmeurer committed:
Add a GitHub icon to the docs footer

asmeurer committed:
Remove background color from blockquotes

asmeurer committed:
Slightly dim the background color of code blocks in light mode

This increases the contrast, while still making the code blocks stand out a
bit on the page.

asmeurer committed:
Factor out the common theme variables again

For whatever reason, the dark theme is not automatically inheriting the light
variables with the latest version of Furo.

asmeurer committed:
Fix the API name color with the latest version of Furo

asmeurer committed:
Fix some theme colors in dark mode

asmeurer commented:
Addressed some comments and additional things I noticed:
- Removed the fancy scrolling when jumping to an anchor (it was too slow on long pages)
- Made the API headings easier to see. They are also now the only part of the page that is colored. 
- "See also" and "Notes" now use a gray background instead of green. "Warning" is still red. 
- Removed the extra vertical space in the parameters section.
- Made the "Parameters", "Returns", "Examples", and so on more prominent. 
- Added a GitHub icon to the footer.
- Made the code block background color slightly dimmer to improve the color contrast. 
- Various small fixes to improve color contrast, especially in dark mode.

I had one idea also, which is to make the function names easier to "spot" on the page by inverting the colors to make the background green (it is currently grey), and the text black (or some other color that's easy to read on green). I'm a little unsure about it, though. Let me know if you think it would be better that way or the way it is now.

asmeurer committed:
Fix "parameters" and "returns" heading formatting for the latest version of Furo

asmeurer commented:
Another note: we are doing enough modifications in custom.css that it might be a good idea to pin the Furo version. We can update it every once in a while, but whenever we do, we have to check if any CSS modifications are required. Even without custom.css, after updating my local version Furo, there were some additional "supported" CSS variables added that required updating.

asmeurer commented:
> I had one idea also, which is to make the function names easier to "spot" on the page by inverting the colors to make the background green (it is currently grey), and the text black (or some other color that's easy to read on green). I'm a little unsure about it, though. Let me know if you think it would be better that way or the way it is now.

Here's a quick mockup of what I mean. Right now, it looks like this:

<img width="760" alt="Screen Shot 2022-03-03 at 5 18 45 PM" src="https://user-images.githubusercontent.com/71486/156675163-a37a1e8c-28ec-44a5-b446-d695cfc78302.png">

I'm wondering if we should instead change it to something like this:

<img width="764" alt="Screen Shot 2022-03-03 at 5 18 34 PM" src="https://user-images.githubusercontent.com/71486/156675184-852d46bc-9a44-402b-8324-5821cf4faffa.png">

Or would that be too noisy on the page? It would only apply to function and class definitions in the API docs, not other headers.

bertiewooster commented:
Thanks for doing all this detailed work, @asmeurer. I do find the second version visually noisy, as I prefer blocks of color to be subtle; that also makes it easier to achieve sufficient color contrast. How about largely keeping the first version, and making the module prefix names text the darker green? The actual function name could be bold to distinguish it from the module prefixes.

asmeurer committed:
Make the logo on the docs site higher resolution

asmeurer committed:
Fix the code line spacing adjustment for the latest version of Furo

asmeurer committed:
Fix various contrast issues in the dark mode docs

asmeurer committed:
Remove duplicate variable from conf.py

asmeurer commented:
Pushed a few more fixes:
- The logo is now high resolution.
- Various contrast fixes for dark mode, especially in the syntax highlighting. 

I'm honestly still not sure how to handle the lighthouse contrast issues in light mode. Unlike dark mode, most of the suggestions for contrast improvements in light mode seem to make the text too close to black to be distinguishable. 

Is there an easy way to run lighthouse on every page, with both light and dark mode? I've been running it one page at a time, but different pages have different elements. There are way too  many pages to run each page manually.

moorepants commented:
I like both, definitely fixes the issues I reaised. Thanks! For the second, the green bars are strong. Maybe a lighter shade of green would be better. The first one is good too though, and sufficient.

bertiewooster commented:
> Is there an easy way to run lighthouse on every page, with both light and dark mode?

[Lighthouse can be run via npm](https://www.npmjs.com/package/lighthouse); I haven't tried that.

I'm not sure if there's a way to automatically do that in both light and dark mode, as the URL doesn't change when I click dark mode. The mode selection does persist when you go from one page to another, so perhaps you could run the Lighthouse scans in light mode, then change to dark mode, then re-running the Lighthouse scans might use the dark mode? I'm not sure if that mode will be picked up by the npm version of Lighthouse.

asmeurer commented:
It looks like you can force dark mode with the `--chrome-flags`. But I don't see an easy way to use this to scan every page. It would be nice if there were a tool that just scanned the local HTML files.

asmeurer committed:
Fix search icon color

asmeurer committed:
Fix some link colors

asmeurer committed:
Clean up some CSS

asmeurer committed:
Fix some indentation

asmeurer commented:
Another thing we can do with Furo is to add a header to each page in the development docs like this:
<img width="1479" alt="Screen Shot 2022-03-04 at 6 30 36 PM" src="https://user-images.githubusercontent.com/71486/156862086-9c821055-ad9d-4393-b65f-c7c2666b5e27.png">

bertiewooster commented:
> Another thing we can do with Furo is to add a header to each page in the development docs like this:

I was thinking it'd be nice to _temporarily_ add headers to each page (in both released and dev docs) when we want to publicize a survey or the like to the community--that would reach anyone who uses the docs, not just people who follow SymPy on Github, Twitter, Gitter, etc. Is there an easy way to do that?

bertiewooster commented:
> I don't see an easy way to use this to scan every page. It would be nice if there were a tool that just scanned the local HTML files.

I think Lighthouse only works if the page is posted on a server. I wonder if there's a way to give it a sitemap to have it visit all the mapped pages?

asmeurer commented:
The announcement feature makes it quite easy to add it to the dev docs. To add it to the "latest" docs we would need to manually rebuild the docs from that git tag with the given header (which isn't too hard).

asmeurer committed:
Make it so the whole sidebar in the docs scrolls

This should hopefully fix the issue with the logo taking up too much space on
mobile.

asmeurer committed:
Mark the current page in the sidebar using a background color

The bold text is too hard to distinguish, and clashes with the bold we are
using for the top-level items.

asmeurer committed:
Fix the link color CSS affecting the sidebar

There's probably a better way to do this, but I don't know enough CSS to know
how.

asmeurer committed:
Add a version selector to the sidebar

This requires a versions.json file to be placed at the root of the SymPy docs
repo like

{
    "latest": "1.10",
    "dev": "1.11.dev"
}

asmeurer committed:
Fix versions.json location

asmeurer committed:
Remove references to the SymPy Live documentation extension

The extension has been removed as part of the move to the Furo theme.

We may eventually want to reenable a similar extension that makes use of
pyiodide. At that point, the text removed here can be reintroduced.

asmeurer commented:
Some updates

- The current page is now highlighted in the sidebar. For example: <img width="725" alt="Screen Shot 2022-03-21 at 6 28 25 PM" src="https://user-images.githubusercontent.com/71486/159384303-b082bde4-266f-4c57-a053-77b9d4c7205b.png">

- I've made it so that the whole sidebar scrolls. Previously the logo would be pinned to the top of the page and only the table of contents would scroll. This should make the sidebar more usable on mobile, where the logo takes up a large amount of vertical space.

- I've added a version selection menu to the sidebar. You can see what it looks like at the bottom of the sidebar at https://www.asmeurer.com/sympy-furo-demo/dev/index.html.
   <img width="292" alt="Screen Shot 2022-03-21 at 6 25 21 PM" src="https://user-images.githubusercontent.com/71486/159384063-f29c8b49-3d01-48da-a2cb-f4bfa066e8b3.png">

   This uses a file called `versions.json` which will be placed at the root of the sympy_doc repo, which looks like

   ```json
   {
       "latest": "1.10",
       "dev": "1.11.dev"
   }
   ```

   If you build the docs locally the file won't be present and the docs will look like this 
   <img width="388" alt="Screen Shot 2022-03-21 at 6 23 17 PM" src="https://user-images.githubusercontent.com/71486/159384029-30d431b1-2ef1-4b01-8cdc-4e1246ce9520.png">

   This is much simpler than the [current solution](https://github.com/sympy/sympy_doc/blob/gh-pages/generate_indexes.py) which injects code into the page after it is built. It also has the advantage that the version selector now appears on every page, not just the index page. Some new code will need to be added to the release process to update the versions.json file. 

- I have removed all references to the SymPy Live extension (mostly from the tutorial). If we ever add a replacement that uses pyiodide, we can reintroduce this text. For now, the extension has been removed because it is not compatible with the Furo theme, and has other issues as well. 

This is now ready for review. I have made all the style changes I wanted to make, so please let me know if you see anything that you feel should be changed in any way. Also if anyone is knowledgeable about CSS or Javascript I would appreciate a review of my CSS in `custom.css` and Javascript in `versions.html`, as I'm not great at either language.

oscarbenjamin commented:
> do you want me to add this to the release scripts or do you want to do it?

It's best if you do this. The relevant place is here:
https://github.com/sympy/sympy/blob/ff5755cecd165cfde49adbe6b4b976dc486fccf3/release/update_docs.py#L60

> This would need to be run whenever either the `latest` or `dev` version changes

Updates to the dev docs are still handled by Travis:
https://github.com/sympy/sympy/blob/ff5755cecd165cfde49adbe6b4b976dc486fccf3/.travis.yml#L121

If we can migrate that to GitHub Actions then we can ditch the Travis scripts as well.

asmeurer commented:
> It's best if you do this. The relevant place is here:

Is the update_docs.py script just run directly by you on the command line or is it called from somewhere else? If I am reading it correctly, we will need to add a new flag to pass in the dev version in addition to the release version (unless there's another way to get that).

oscarbenjamin commented:
> Is the update_docs.py script just run directly by you on the command line or is it called from somewhere else?

I run it directly. The sequence is:

1. Push tag and wait for CI
2. Download release files from CI
3. Run `release/github_release.py`
4. Run `twine upload`
5. Run `release/update_docs.py`
6. Edit release notes
7. Send an email to the mailing list.

It's fine to pass in the dev version manually to the update_docs.py script.

asmeurer committed:
Update the update_docs script for the new versions.json process

asmeurer committed:
Remove the generate_indexes.py command from Travis

asmeurer commented:
I updated the release script. It's kind of hard to test, so you'll have to let me know if everything works properly when we do the next release. For now, I have added `versions.json` to the doc repo manually. I will delete the `generate_indexes.py` script there once this is merged.

oscarbenjamin commented:
> It's kind of hard to test, so you'll have to let me know if everything works properly when we do the next release.

We can check when that happens. To be clear it's not that hard to test though:

1. Clone the docs repo as ../sympy_doc
2. Checkout your PR branch
3. Change the version in sympy/release.py to 1.11
4. Run `release/build_docs.py 1.11 .`
5. Run `release/update_docs.py ../sympy_doc ./sympy-docs-xxx.zip 1.11`

If you **don't** pass `--push` to update_docs.py it will do everything except for the final push so you can open `index.html` and check how it looks.

asmeurer committed:
Fix the color of the "Hide Search Matches" text

asmeurer committed:
Fix the update_docs script

asmeurer commented:
OK, I was able to get it working (and fixed a bug with it). Note that the last line will now be `release/update_docs.py ../sympy_doc ./sympy-docs-xxx.zip 1.11 1.12.dev`.

Also note that if you just open `index.html` in your browser, it will say "other versions not found". This is because browsers' cross-origin policies will not allow Javascript to load the versions.json file when using a `file://` URL. If you instead use a web server, like `python -m http.server` it will load the versions correctly (and it will obviously work correctly on the real docs.sympy.org site).

bertiewooster commented:
Coming together nicely!

## Color issues

- Contrast is a bit low for links of regular-sized text. (This is the only accessibility issue I've found using Lighthouse, giving pages a score of 98/100, very good.)
- Bright green text color (h2's, links, etc.) strikes me as a little garish. 

Possible solution to both: Could the green in the body text (h2 and links) be the dark green, and the green background of the navbar be a lighter green? (This may have already been tried and found unsatisfactory?)

## Content
The following are sort of tangential, could be a separate pull request: On homepage, suggest renaming top-level categories, and making them the same in the body and in the left navbar:

- SymPy Tutorials -> Tutorials
- SymPy Guides -> How-to Guides (only left navbar name to be changed)
- SymPy API Reference / Reference Documentation -> API Reference

Similarly, moving pages from Miscellaneous into one of the existing Diataxis categories, probably Guides.

asmeurer commented:
> Nitpick: seems like there are two different green link colors, for example on [style guide](https://www.asmeurer.com/sympy-furo-demo/dev/guides/contributing/documentation-style-guide.html), for example link in sphinx.ext.linkcode seems lighter green, and link in numpydoc seems darker green.

The darker green means the link has been visited. Maybe this is too subtle? I've noticed that many websites these days don't even distinguish these any more (in the old days, unvisited links were blue and visited links were purple). Note that our current docs theme does this as well. 

> Contrast is a bit low for links of regular-sized text. (This is the only accessibility issue I've found using Lighthouse, giving pages a score of 98/100, very good.)

My issue with this has been making sure the contrast of the link color is itself high enough relative to the surrounding text. If you let the Chrome dev tools suggest a better contrast color, it will be easier to read against the background, but will look almost indistinguishable from the surrounding black text. That's bad because you want to be able to tell that a link is a link. 

Here's the current style with no links visited

<img width="751" alt="Screen Shot 2022-03-24 at 4 44 34 PM" src="https://user-images.githubusercontent.com/71486/160022326-421db22b-7be9-4566-836c-1e06d5ee91c3.png">

Here's the current style with the first link visited

<img width="753" alt="Screen Shot 2022-03-24 at 4 45 13 PM" src="https://user-images.githubusercontent.com/71486/160022348-3df8df6c-6a60-4ec0-836f-e8b469bf4fce.png">

And here's the current style with all links visited

<img width="751" alt="Screen Shot 2022-03-24 at 4 44 11 PM" src="https://user-images.githubusercontent.com/71486/160022399-4d9fe126-9075-4bff-8150-f433b174fdcd.png">

Perhaps we can just use latter for all links, and have no distinction between visited and unvisited. 

Personally, I prefer the current style, but I also don't have issues seeing color (both are readable to me), so I'm not sure just how bad the contrast issue of the current link color is. 

Just to be clear, the current link color is #29A329, which has a ratio of [3.3](https://whocanuse.com/?b=ffffff&c=29a329&f=16&s=) on a white background and the current visited link color is #307748, which has a ratio of [5.44](https://whocanuse.com/?b=ffffff&c=307748&f=16&s=) on a white background. AA is 4 and AAA is 7. I've found that AAA is basically untenable because it makes the links indistinguishable from the surrounding black text.

In dark mode, #307748 has lower contrast in dark mode ([3.86](https://whocanuse.com/?b=000000&c=307748&f=16&s=)), whereas #29A329 has good contrast ([6.37](https://whocanuse.com/?b=000000&c=29a329&f=16&s=)). Here, Chrome's suggested 7 contrast ratio color [#2FB92F](https://whocanuse.com/?b=000000&c=2fb92f&f=16&s=) is actually distinghable from the surrounding white text, at least to my eyes. 

I just found https://webaim.org/resources/linkcontrastchecker/, which actually lets you check link text against the surrounding text and the background text. According to it, using [#29A329 for dark mode](https://webaim.org/resources/linkcontrastchecker/?fcolor=FFFFFF&bcolor=000000&lcolor=29A329) and [#307748 for light mode](https://webaim.org/resources/linkcontrastchecker/?fcolor=000000&bcolor=FFFFFF&lcolor=307748) are both good contrast against both the background and the surrounding text (AA). It also correctly identifies that Chrome's "contrast ratio 7" suggestions have too poor contrast for the link text against the surrounding text. For instance, Chrome suggests [#2FB92F](https://webaim.org/resources/linkcontrastchecker/?fcolor=FFFFFF&bcolor=000000&lcolor=2FB92F) for dark mode, which is too low contrast against white text, and [#1A661A](https://webaim.org/resources/linkcontrastchecker/?fcolor=000000&bcolor=FFFFFF&lcolor=1A661A) for light mode, which is too low contrast against black text.

Dark mode is going to have better contrast in general, since I also adjusted the syntax highlighting colors to be better in dark mode.

**So tl;dr: my suggestion is to ditch the current visited/unvisited distinction for link colors and use the current darker "visited" color in light mode and the current lighter "unvisited" color in dark mode.** If we want to keep the visited/unvisited distinction we will need to think harder about this. 

> Bright green text color (h2's, links, etc.) strikes me as a little garish.

Regarding the green h2, this is actually only the case on the main page. It's because on that page, [the headers are links](https://github.com/sympy/sympy/blob/master/doc/src/index.rst). As I recall, we did this so that the sections are very evident as clickable things on the page, while still being part of the top-level table of contents. If you have any suggestions on a better way to do this, I'm open to it. I don't think it's a perfect solution because this also makes them headers with anchors, which is confusing. I don't think we should change the styling of links everywhere else just because this one thing happens to be rendering as a link right now. 

We actually have some pretty inconsistent use of headers throughout the documentation, which Furo is going to make more obvious. One example is that sometimes we have a header per reference function, like at https://www.asmeurer.com/sympy-furo-demo/dev/modules/core.html, which produces a list of all the functions in the sidebar. In other cases we only have top-level headers, like at https://www.asmeurer.com/sympy-furo-demo/dev/modules/ntheory.html. We should probably prefer the former style since it produces the nice list in the sidebar. But it's not trivial: the latter style generally comes from using `automodule` rather than explicitly listing each function with `autofunction` in the `rst`, so if we switch to always explicitly listing each function, we need to make better use of things like the `bin/coverage_doctest.py` script to make sure we don't omit functions from Sphinx. Alternately, we might be able to modify autodoc so that `automodule` automatically produces headers (this might be possible by messing with autodoc templates, I'm not sure). 

There are other problems too. For example, if you click on "number theory" on the sidebar, it takes you to this page, which is kind of useless: https://www.asmeurer.com/sympy-furo-demo/dev/reference/public/numbertheory/index.html. You have to click on the headers under it to go to the actual page on number theory. But those headers should really be on the right sidebar, not the left one. This is an issue with the way the rst pages and tables of contents are laid out for the number theory page, not with Furo.

> Possible solution to both: Could the green in the body text (h2 and links) be the dark green, and the green background of the navbar be a lighter green? (This may have already been tried and found unsatisfactory?)

Lighter green in the sidebar definitely does not look good. There are actually two lighter green colors we are using, and both are too bright for a background color.

<img width="1469" alt="Screen Shot 2022-03-24 at 4 35 45 PM" src="https://user-images.githubusercontent.com/71486/160021358-29c2bc8d-aeee-44b8-8499-ab62bb3cbfa8.png">
<img width="1464" alt="Screen Shot 2022-03-24 at 4 35 34 PM" src="https://user-images.githubusercontent.com/71486/160021362-1ce2495e-fae5-4241-a6c6-3977a6df8518.png">

And it's obviously even worse with dark mode. It's actually kind of nice with the current background color that it looks good in both light and dark mode. The current sidebar green is dark enough to be unobtrusive but also gives the SymPy docs its flair.

> The following are sort of tangential, could be a separate pull request: On homepage, suggest renaming top-level categories, and making them the same in the body and in the left navbar:

The content is independent of the theme. Let's make those changes in a separate PR. They can be made before or after this is merged.

bertiewooster commented:
> I just found https://webaim.org/resources/linkcontrastchecker/, which actually lets you check link text against the surrounding text and the background text.

Great resource. Thanks!



> So tl;dr: my suggestion is to ditch the current visited/unvisited distinction for link colors and use the current darker "visited" color in light mode and the current lighter "unvisited" color in dark mode.

I concur.

I'll work on a pull request for changing top-level category names, and moving Miscellaneous into Guides.

asmeurer committed:
Make the header text white

The gray just looks kind of bad.

asmeurer committed:
Use better contrast colors for links

This removes the separate colors for visited and unvisited. We now use the
darker of the two in light mode and the ligher of the two in dark mode. These
colors satisfy AA color contrast, both against the background and against the
surrounding black/white text (see
https://github.com/sympy/sympy/pull/23159/#issuecomment-1078480215).

This replaces the "brand-content" color in light mode since this is primarily
used for links. We still use the lighter color for the API function names in
light mode, so that they pop out more (dark mode already has a lighter color
for API function names).

asmeurer committed:
Remove redundant CSS variable

asmeurer committed:
Remove the green underline on "SymPy documentation" on hover

asmeurer commented:
I've made the suggested changes to link colors.

bertiewooster commented:
@asmeurer thanks for the changing the link colors so they're all dark, regardless of whether they've been visited or not.

I ran [Lighthouse](https://googlechrome.github.io/lighthouse/viewer/?psiurl=https%3A%2F%2Fwww.asmeurer.com%2Fsympy-furo-demo%2Fdev%2Fguides%2Fcontributing%2Fdocumentation-style-guide.html&strategy=desktop&category=performance&category=accessibility&category=best-practices&category=seo&category=pwa&utm_source=lh-chrome-ext) again and found a couple issues:

- The only remaining accessibility contrast issue I found is the bluish text in a code block, for example the class name and docstring (below, "loggamma" and "r""" The loggamma function...". My naive suggestion would be to decrease the saturation (intensity of color) of the code-block background, that is ["tint"](https://en.wikipedia.org/wiki/Tint%2C_shade_and_tone) it. (Personally, I find blocks of intense color distracting anyway.)
<img width="583" alt="Screen Shot 2022-03-25 at 10 08 44 PM" src="https://user-images.githubusercontent.com/1335522/160220671-da9bd77f-ae6f-4979-87ac-3d7f756e4cc1.png">
- In terms of search engine optimization (SEO), one thing the pages are missing is a meta description: "Meta descriptions may be included in search results to concisely summarize page content." I learned that [Sphinx can support meta descriptions](https://docs.readthedocs.io/en/stable/guides/technical-docs-seo-guide.html) using reStructuredText like

```
.. meta::
    :description lang=en:
```

so I'm guessing the issue is those descriptions aren't auto-generated? I suppose we could manually generate descriptions for the non-API pages...

asmeurer commented:
> The only remaining accessibility contrast issue I found is the bluish text in a code block, for example the class name and docstring (below, "loggamma" and "r""" The loggamma function...". My naive suggestion would be to decrease the saturation (intensity of color) of the code-block background, that is ["tint"](https://en.wikipedia.org/wiki/Tint%2C_shade_and_tone) it. (Personally, I find blocks of intense color distracting, so this is my favored solution. :-) )

I already have decreased intensity of the code background color (in light mode). It was originally the same color as it is in the current site. 

I had also adjusted some of the code highlighting colors in dark mode to be better contrast. But I'm wondering if instead we ought to adjust the pygments theme, which is separate from the docs theme. Right now it's set to "sphinx" which is designed around the original docs theme, and probably wasn't designed with contrast in mind as much as a more modern one would be. It's actually the same theme used by the official Python docs (see e.g., https://docs.python.org/3/tutorial/introduction.html). It just happens to work well with the green background, so we never changed it. 

I'll see what the other options are. In either case, I'm realizing now that adjusting the syntax highlighting via CSS is the wrong approach. It would be better to adjust the pygments theme itself. 

I will say that I do like having the code blocks stand out somewhat. The doctest examples are the most important part of the docstring, so having them be prominent is a good thing. 

> In terms of search engine optimization (SEO), one thing the pages are missing is a meta description: "Meta descriptions may be included in search results to concisely summarize page content." I learned that [Sphinx can support meta descriptions](https://docs.readthedocs.io/en/stable/guides/technical-docs-seo-guide.html) using reStructuredText like

Does this need to be added to each page manually, or is the meta description something that can be generated automatically? 

I was considering enabling https://github.com/wpilibsuite/sphinxext-opengraph. I'm not sure if this is related to this or if it's something else. 

Another thing I noticed that Furo is missing compared to the current site is the `rel="canonical"` tag (see https://github.com/sympy/sympy/blob/master/doc/src/_templates/layout.html), although to be honest, it wasn't clear to me that that ever actually did much to the Google results.

bertiewooster commented:
> The [Open Graph protocol](https://ogp.me/) enables any web page to become a rich object in a social graph. For instance, this is used on Facebook to allow any web page to have the same functionality as any other object on Facebook.
>[Basic Metadata](https://ogp.me/#metadata)
`<meta property="og:title" content="The Rock" />`

My guess is that these meta properties do not apply to search engines; this is the first I've learned about it, so I could be wrong.

> > In terms of search engine optimization (SEO), one thing the pages are missing is a meta description: "Meta descriptions may be included in search results to concisely summarize page content." I learned that [Sphinx can support meta descriptions](https://docs.readthedocs.io/en/stable/guides/technical-docs-seo-guide.html) using reStructuredText like
> 
> Does this need to be added to each page manually, or is the meta description something that can be generated automatically?

From [Read The Docs](https://docs.readthedocs.io/en/stable/guides/technical-docs-seo-guide.html#use-meta-tags):
>Typically search engines will use the first few sentences of a page if no meta description is provided.

So if we don't supply anything, search engines will extract some text from the page. As these Google search results for "sympy solve" demonstrate, that may not be the most helpful text for a searcher to decide if the page is relevant:
<img width="686" alt="Screen Shot 2022-03-28 at 5 46 59 PM" src="https://user-images.githubusercontent.com/1335522/160492848-9cf9fdff-435c-4969-b69a-3e7cc005810c.png">

Here's the guidance for a meta description:
> - Your meta description should have the most relevant content of the page. A searcher should know whether they’ve found the right page from the description.
> - The meta description should be between 150-300 characters and it may be truncated down to around 150 characters in some situations.
> - Meta descriptions are used for display but not for ranking.

So to improve on what's automatically extracted, I think the descriptions would need to be generated manually. Having said that, arguably the meta description should also be the first (or nearly) text that the visitor encounters, at the top of the page. For example, on our [Matrices page](https://docs.sympy.org/dev/modules/matrices/matrices.html), the first text the visitor encounters is:

> ## Matrices (linear algebra)
> ### Creating Matrices
> The linear algebra module is designed to be as simple as possible. First, we import and declare our first Matrix object:

It would definitely benefit from a better overview of the topic, which could be the same as the meta description.

So perhaps we should update the documentation style guide to include a 1-2 sentence overview for each page, which would also be used as the meta description. I haven't investigated whether there's some way to automate tagging that overview as the meta description.

asmeurer committed:
Remove unused layout.html template

asmeurer committed:
Add html_basename to the docs

This should add rel="canonical" links to each page, except it doesn't actually
seem to work.

asmeurer commented:
The search engine thing is specifically the rel="canonical". See https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls. It theoretically would make it so that Google would prefer "latest" URLs over "dev" URLs. It also should make it so that Google doesn't try to serve versions of the docs that people have built and published themselves (for instance, my demo site). In practice, it's not perfect in my experience. We used to serve many old versions of the docs, and Google would generally serve up a random old URL rather than the latest one, even with the rel="canonical" (see https://github.com/sympy/sympy_doc/pull/33). Even so, if it does help even a little bit it's worth having there. 

Apparently rel="canonical" is supported directly in Sphinx https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_baseurl. So I've added that, although I think it doesn't actually work currently because of a bug in Furo https://github.com/pradyunsg/furo/discussions/406.

For the opengraph and meta description stuff, let's do it in a separate pull request. It's basically unrelated to the theme, and doesn't exist on the current site so it's not a regression. 

For the code highlighting, it looks like the "sphinx" style we have been using is basically the [friendly](https://pygments.org/styles/) pygments style with a [yellow background](https://github.com/sphinx-doc/sphinx/blob/b12b39db05d605cfb946f84afd27e0f6cf3c9e13/sphinx/pygments_styles.py). 

For dark mode, Furo is defaulting to monokai. The [Pygments styles page](https://pygments.org/styles/) says that both friendly and monokai are lower contrast. 

I'm not really sure what I want to do for the syntax highlighting. I'm not really keen on building a custom syntax highlighting theme just for the SymPy docs. So I would suggest trying to find an existing theme that works well. Or just using the Furo defaults. I'm not sure if pygments/sphinx would accept upstream improvements to improve the contrast of their themes, or if they would consider changing the theme colors to be a backwards incompatibility break.

asmeurer committed:
Use custom pygments styles with better color contrast in the docs

We use the same styles as before (sphinx for light mode and the Furo default
of native for dark mode), but they have been updated so that any text colors
that were lower than 4.5 color contrast have been modified. The styles mostly
look the same but some parts may be easier to read. The 4.5 color contrast
makes the docs WCAG AA compliant.

asmeurer committed:
Fix some wording

asmeurer committed:
Remove the pre line-height from custom.css

This is now set as part of the pygments themes.

asmeurer commented:
For the syntax highlighting, I found this https://github.com/mpchadwick/pygments-high-contrast-stylesheets, which has a tool that allows updating pygments styles to be WCAG AA compliant. I used this to create versions of the two themes that have better contrast. I outlined the process I used in [this gist](https://gist.github.com/asmeurer/d78409ea033aeeb764a11bb4acaeccc2). It's a bit hacky, and ideally the whole thing would just be one Python script, but since I don't plan on doing this again, I am OK with the hacky process.

This also reverts back to the original darker yellow background for the light mode, since the lighter yellow didn't even work for better contrast (it actually made things like the blue docstrings worse), and I like the darker yellow anyway (it highlights the code blocks more, and it fits well with the green theme). 

We might want to try upstreaming these changes, although I'm unsure if pygments or sphinx would accept color changes to their themes or if they would consider that to be backwards incompatible. @pradyunsg you may also be interested in this if you want Furo to have better WCAG compatibility. In particular, the default style you use for dark mode, "native", has some color contrast issues that are fixed here. I didn't check if the Furo default style for light mode has any such issues. If you're interested in upstreaming any of the other custom.css changes I have here or the versions menu, I would be open to that as well.

I think this should be ready to go now. The rel="canonical" thing will need to be fixed upstream in Furo, and the meta descriptions and opengraph stuff can be done separately, as can the modifications to the text and the headings layout.

bertiewooster commented:
Thanks @asmeurer. Is the version posted at https://www.asmeurer.com/sympy-furo-demo/dev/ the latest version?

asmeurer commented:
Yes it is.

pradyunsg commented:
You likely want to override `--color-sidebar-background-border` as well.

pradyunsg commented:
> And I would need to figure out how to make it cross-link with Javascript since it doesn't seem to be Jinja templated.

You can use `{% block announcement %}` in page.html, to provide the internal content of the announcement, if that's what you want -- as long as you set `announcement` in `html_theme_options` to a truthy value.

I should probaby document how that works.

pradyunsg commented on doc/src/_static/custom.css:
```diff
@@ -0,0 +1,116 @@
+/* Completely disable blockquotes. The Furo defaults make block quotes very*/
+/* apparent. However, the parameter blocks from numpydoc render as block
+ * quotes, which looks bad. Furthermore, it is very easy for things to
+ * accidentally have block quotes due to the fact that RST makes indented
+ * blocks render as a block quote (unlike Markdown which makes indented blocks
+ * render as code). There aren't any places in the documentation that use a
```
Well, this is your documentation, so do whateverr you want -- I'll note that I made blockquotes apparent *because* it was too easy to inject them by accident with reST, making them unusable when you *actually* want them. :)

asmeurer replied:
I could make an effort to remove all the blockquotes, and maybe I should. This is a much easier thing for now. A big problem is that numpydoc is using blockquote for the parameters, which makes them look pretty bad. So we'd need to fix that (we're actually currently using a forked version of numpydoc, which could be part of the problem, but that just highlights the complexity). SymPy has a lot of documentation (over 270 RST files, most of which are pulling in documentation from docstrings). So any kind of content clean up like this would take a bit of effort, which is more than what I want to do for this PR.

pradyunsg commented on doc/src/_static/custom.css:
```diff
@@ -0,0 +1,116 @@
+    color: white !important;
+}
+
+/* Make the "class", "property" and so on before function names in the API*/
+/* docs bold. */
+em.property:first-child {
+    font-weight: bold;
+}
```
The entire signature is bold by default, so you shouldn't need this.

asmeurer replied:
Looks like you are right. This must have been left over from a previous iteration.

pradyunsg commented on doc/src/_templates/sidebar/versions.html:
```diff
@@ -0,0 +1,56 @@
+}
+-->
+<div class="sidebar-tree sidebar-versions">
+  <h5 class="versions-header">Documentation Version</h4>
```
I would suggest adding a padding-left to this, to align it with the rest of the sidebar content. You might want to consider if you want this to look like sidebar captions instead of a heading.

asmeurer replied:
I'll have to play with it some more. I'm OK with the way it currently looks but I'll see if it looks better this way. Note that I'm not a designer and my HTML/Javascript/CSS skills are terrible, so if you see something that looks off, it's most likely unintentional so please let me know.

asmeurer replied:
Actually this was supposed to have the exact same `padding` as the rest of the sidebar, but my editor inserted an incorrect line break in the CSS and messed it up. Should look better now.

pradyunsg commented on doc/src/conf.py:
```diff
@@ -142,12 +146,96 @@
+    "light_css_variables": common_theme_variables,
+    # The dark variables automatically inherit values from the light variables
+    "dark_css_variables": {
+        **common_theme_variables,
```
You likely do not need this.

asmeurer replied:
If I comment this line out, the light specific modifications do not carry over to dark mode. I think it was working before when I had an older version of Furo but it stopped working when I updated it.

pradyunsg replied:
Hmm... Fascinating. I'll dig into what's happening here later. I'm pretty sure this won't do the wrong thing though, so this is definitely fine to do anyway. :)

pradyunsg commented on doc/src/conf.py:
```diff
@@ -220,6 +308,7 @@
 # SymPy logo on title page
 html_logo = '_static/sympylogo.png'
 latex_logo = '_static/sympylogo_big.png'
+html_favicon = '../_build/logo/sympy-notailtext-favicon.ico'
```
This should likely be logo/... instead? Otherwise, you won't get the logo on the first build (unless I'm misunderstanding your build process).

asmeurer replied:
Yeah, we compile the logo from svg to png and ico as part of the build process (the Makefile dependencies ensure it's compiled first).

pradyunsg commented:
> We might want to try upstreaming these changes, although I'm unsure if pygments or sphinx would accept color changes to their themes or if they would consider that to be backwards incompatible. @pradyunsg you may also be interested in this if you want Furo to have better WCAG compatibility.

If pygments gets new better-contrast themes or improves their themes, I'd be *very* happy to see that. I had considered writing custom pygments theme that had a better contrast story, but I decided against that to minimise how much stuff I was doing in the theme; to instead try contributing to those project.

I haven't actually had the time/mental space to come around to doing that though, so I welcome this! :)

asmeurer commented:
Thanks for the review @pradyunsg.

> You likely want to override --color-sidebar-background-border as well.

Wow that's a very subtle thing, especially with my other overrides. I guess the best thing to set this to when the sidebar has a different background color as in our case is to `var(--color-background-primary)`.

> You can use {% block announcement %} in page.html, to provide the internal content of the announcement, if that's what you want -- as long as you set announcement in html_theme_options to a truthy value.

I don't think we have any announcements we want to add at this point, but that's good to know for the future. We could consider adding one for the dev pages, but I still unsure if it's a good idea.

> If pygments gets new better-contrast themes or improves their themes, I'd be very happy to see that. I had considered writing custom pygments theme that had a better contrast story, but I decided against that to minimise how much stuff I was doing in the theme; to instead try contributing to those projects instead.

I more or less feel the same way. I didn't want to spend a whole lot of time on this, as I know it could end up being a huge rabbit hole that would ultimately produce a lot of good work, but not end up being related to the SymPy documentation. That's why I came up with a pretty hacky solution that just reuses the Ruby script I found at https://github.com/mpchadwick/pygments-high-contrast-stylesheets/blob/gh-pages/tools/make-stylesheet. It gets the job done for us, and makse the SymPy docs WCAG compatible.

Ideally the whole thing would be done in Python (I'm not sure what the best library for that would be), and and be more convenient to use. For instance, one could have a function that just takes a pygments theme and automatically makes it WCAG AA compatible without requiring any manual. Or alternately, just change the themes upstream to be AA compatible. But given that pygments [specifically notates](https://pygments.org/styles/) which styles have bad contrast gives me the impression that simply updating them is not a simple thing for them to do. The "sphinx" style is actually part of Sphinx, so updating that would be a separate discussion (but again, likely not a simple proposition, as it would affect a lot of projects, including docs.python.org itself).

So for now I'm going to leave this as an open project for someone else to implement, but if you or anyone else ever does, please let me know.

bertiewooster commented:
The [Lighthouse report](https://googlechrome.github.io/lighthouse/viewer/?psiurl=https%3A%2F%2Fwww.asmeurer.com%2Fsympy-furo-demo%2Fdev%2Fmodules%2Fntheory.html&strategy=desktop&category=performance&category=accessibility&category=best-practices&category=seo&category=pwa&utm_source=lh-chrome-ext) reveals an accessibility contrast issue with class etc. green names against a gray background, whether the green text is [regular](https://www.asmeurer.com/sympy-furo-demo/dev/modules/ntheory.html#sympy.ntheory.generate.Sieve) or [bold](https://www.asmeurer.com/sympy-furo-demo/dev/modules/ntheory.html#sympy.ntheory.generate.Sieve.extend):

<img width="749" alt="sympy-docs-contrast-issue-gray-background" src="https://user-images.githubusercontent.com/1335522/161443453-ddd40991-49a8-4fbe-9400-053778b9a4e2.png">

That situation seems to be the only contrast issue; the code blocks are fine.

asmeurer committed:
Remove an unnecessary piece of CSS

asmeurer committed:
Fix the sidebar border color

asmeurer committed:
Make the API function names have better color contrast

asmeurer committed:
Tone down the brand-primary color, and use it for the sidebar hover

It's only used in two places, the left sidebar hover, and the right sidebar
active page. This color was taken from the SymPy logo, and modified slightly
to have a 4.5 contrast ratio. Unlike the previous color, this one now has good
contrast, except in dark mode, where we use the old color for the black
background.

asmeurer committed:
Clean up the dark theme colors

- Use color-brand-content and color-brand-primary instead of redefining
  specific things.
- Remove the gray selected header color. It creates bad color contrast. The
  default yellow is easier to work with.
- The primary and content colors now have at least 4.5 contrast, including
  when the header is selected.
- Remove some colors that were defined to nearly identical colors as the
  defaults (but which still have fine color contrast).

asmeurer committed:
Reapply the color-foreground-secondary for dark mode

Apparently it actually is needed for color contrast compared to the default.

asmeurer committed:
Fix the sidebar hover colors in dark mode

asmeurer commented:
I fixed that and cleaned up some of the colors a bit (we now just have the two greens in color-brand-content and color-brand-primary). Also fixed some contrast issues that only showed when selecting a header (lighthouse does not see these unless you give a URL with an anchor). I also changed the hover color in the left sidebar because it had bad contrast (lighthouse does not look at hover colors).

asmeurer committed:
Fix incorrect line wrapping in the CSS

bertiewooster commented:
Great, thanks for going beyond even the automated Lighthouse checks to ensure that as many people as possible can use SymPy documentation.

Comparing the current version (above) to the restyled version (below) demonstrates how much cleaner, more navigable, and more functional the restyled site will be:

<img width="1134" alt="sympy-current-example" src="https://user-images.githubusercontent.com/1335522/162097237-e0042652-59f9-4cd0-9ef8-3e332105e864.png">
<img width="1134" alt="sympy-furo-example" src="https://user-images.githubusercontent.com/1335522/162097238-db23b4a0-fde3-490b-9ec7-c36e31bf0de6.png">
.

asmeurer commented:
If there are no further comments, I'd like to merge this. If we find something later that I've missed, we can make additional PRs to fix. It will be good to have this in the dev docs soon so that we can find any such issues before it gets released to the main "latest" documentation. 

Let's open new issues for the items we discussed above that were not addressed in this PR.

oscarbenjamin commented:
Looks good to me. Let's get this in.

asmeurer commented:
Looks like I forgot to update the dependencies for Travis. Fix forthcoming.

asmeurer commented:
I opened several issues for things that were discussed here but not addressed (see the cross references above this comment). If I missed anything, please open an issue for it.
